packages/blunderbuss/src/blunderbuss.ts

```diff
@@ -129,6 +133,13 @@ export function blunderbuss(app: Probot) {
       const byConfig = issue ? config.assign_issues_by : config.assign_prs_by;
       const issuePayload = issue || pullRequest;
 
+      // Acquire the lock.
+      const lock = new DatastoreLock('blunderbuss', lockTarget);
+      const lockResult = await lock.acquire();
+      if (!lockResult) {
+        throw new Error('Failed to acquire a lock');
+      }
+
```

Comment:
Once the lock is successfully acquired, we need to refresh the issue's data (to ensure it hasn't changed outside of the critical section). We may also want to introduce a brief sleep (500ms to 1s) to give auto label a chance to do its thing.

Reply:
You're right in a long run, but I think it is better to introduce the lock first, then do another round of refactoring.

For giving auto label a chance, we already have 10 seconds sleep, and now it does the trick.

Reply:
It's done here:
https://github.com/googleapis/repo-automation-bots/blob/819e646055cfd277f673a3caa80cf1c2d49536fb/packages/blunderbuss/src/blunderbuss.ts#L195

Reply:
I don't think it will fix #1171, because the lock itself isn't enough. While it prevents blunderbuss from operating on any individual issue concurrently, it's still operating on an outdated copy of that issue's information that was sent in the webhook. Imagine the following scenario (where T(X) is an action taking place at some arbitrary X unit of time):

T(0): Someone opens a PR and a webhook event is triggered (Issue = no assignees)
T(1): Blunderbuss receives the event from T(0) and sleeps for 10
T(2): The issue is applied a label and another webhook event is triggered
T(1): Blunderbuss receives the event from T(2) and sleeps for 10
T(11): Blunderbuss process the event at T(0), acquires the lock, assigns it to someone
T(12): Blunderbuss process the event at T(1), acquires the lock, assigns it to someone else

If the issue status is refreshed after the lock is acquired, Blunderbuss will know it has assignees and will not reassign (unless it's the `blunderbuss: assign` label, which allows for assigning to multiple times).

Reply:
(in fact with the introduction of the lock and the updating of issues, we can probably significantly reduce the sleep time without experiencing the race condition - 10s is a long time to wait)

Reply:
(If you still want to wait to a separate PR to do this, let's just make sure we don't close #1171 by merging this one)

Reply:
Ya, separate PR. I changed the comment to `part of #1171`